File: sdks/local-rp/csharp/src/LinkKeys.LocalRp/Dns/DnsParseError.cs
namespace LinkKeys.LocalRp.Dns;

/// <summary>A <c>_linkkeys</c>/<c>_linkkeys_apis</c> TXT record failed to parse. Mirrors <c>liblinkkeys::dns::DnsParseError</c>.</summary>
public class DnsParseError : Exception
{
    public enum ErrorKind
    {
        NoLinkKeysRecord,
        MissingVersion,
        UnsupportedVersion,
        MissingApisEndpoint,
        InvalidFormat,
    }

    public ErrorKind Kind { get; }

    public DnsParseError(ErrorKind kind, string? detail) : base(detail is null ? kind.ToString() : $"{kind}: {detail}")
    {
        Kind = kind;
    }

    /// <summary>The symbolic string <c>dns.json</c>'s <c>expected_error</c> field uses.</summary>
    public string Symbol => Kind switch
    {
        ErrorKind.NoLinkKeysRecord => "no_linkkeys_record",
        ErrorKind.MissingVersion => "missing_version",
        ErrorKind.UnsupportedVersion => "unsupported_version",
        ErrorKind.MissingApisEndpoint => "missing_apis_endpoint",
        ErrorKind.InvalidFormat => "invalid_format",
        _ => throw new ArgumentOutOfRangeException(),
    };
}

File: sdks/local-rp/java/src/main/java/community/catalyst/linkkeys/localrp/dns/DnsParseError.java
package community.catalyst.linkkeys.localrp.dns;

/** A {@code _linkkeys}/{@code _linkkeys_apis} TXT record failed to parse. Mirrors {@code liblinkkeys::dns::DnsParseError}. */
public class DnsParseError extends RuntimeException {
    public enum Kind {
        NO_LINKKEYS_RECORD,
        MISSING_VERSION,
        UNSUPPORTED_VERSION,
        MISSING_APIS_ENDPOINT,
        INVALID_FORMAT,
    }

    private final Kind kind;

    public DnsParseError(Kind kind, String detail) {
        super(detail == null ? kind.toString() : kind + ": " + detail);
        this.kind = kind;
    }

    public Kind kind() {
        return kind;
    }

    /** The symbolic string {@code dns.json}'s {@code expected_error} field uses. */
    public String symbol() {
        return switch (kind) {
            case NO_LINKKEYS_RECORD -> "no_linkkeys_record";
            case MISSING_VERSION -> "missing_version";
            case UNSUPPORTED_VERSION -> "unsupported_version";
            case MISSING_APIS_ENDPOINT -> "missing_apis_endpoint";
            case INVALID_FORMAT -> "invalid_format";
        };
    }
}
